tool/gravity/cli/checks.go

```diff
@@ -17,37 +17,89 @@ limitations under the License.
 package cli
 
 import (
+	"context"
 	"fmt"
 	"io/ioutil"
+	"time"
 
+	"github.com/gravitational/gravity/lib/app"
 	"github.com/gravitational/gravity/lib/checks"
+	"github.com/gravitational/gravity/lib/constants"
+	"github.com/gravitational/gravity/lib/fsm"
 	"github.com/gravitational/gravity/lib/localenv"
+	"github.com/gravitational/gravity/lib/pack"
 	"github.com/gravitational/gravity/lib/schema"
+	statusapi "github.com/gravitational/gravity/lib/status"
+	upgradechecks "github.com/gravitational/gravity/lib/update/cluster/checks"
 
+	"github.com/fatih/color"
 	pb "github.com/gravitational/satellite/agent/proto/agentpb"
 	"github.com/gravitational/trace"
 )
 
-func checkManifest(env *localenv.LocalEnvironment, manifestPath, profileName string, autoFix bool) error {
-	data, err := ioutil.ReadFile(manifestPath)
+type preflightChecksConfig struct {
+	manifestPath string
+	imagePath    string
+	profileName  string
+	autoFix      bool
+	timeout      time.Duration
+}
+
+func executePreflightChecks(env *localenv.LocalEnvironment, config preflightChecksConfig) error {
+	ctx, cancel := context.WithTimeout(context.Background(), config.timeout)
+	defer cancel()
+
+	if !detectCluster(ctx) {
+		env.PrintStep("No deployed cluster detected, running install preflight checks")
+		return checkInstall(ctx, env, config)
+	}
+
+	env.PrintStep("Detected deployed cluster, running upgrade preflight checks")
+	return checkUpgrade(ctx, env, config)
+}
+
+// detectsCluster attempts to detect whether the current node has a cluster deployed.
+//
+// Returns true if the cluster is detected and false otherwise.
+func detectCluster(ctx context.Context) bool {
+	_, err := statusapi.FromPlanetAgent(ctx, nil)
```

Comment:
Hmm, I'm not sure this logic should be based on reaching a service, doesn't it make sense to detect the cluster based on whether the cluster has been installed to disk, and not based on the status/health of a service?

Reply:
Yeah, we were discussing it a bit with Dima above as well, and as I mentioned I don't think we have the "silver bullet" way to determine whether the cluster is installed right now.

For example, the way you mentioned also won't work for scenarios where, say, a cluster was partially installed/uninstalled and still has some leftovers on disk.

At least by checking the planet agent status we know that if it succeeds, the cluster **for sure** is deployed. If it doesn't succeed, yeah, it may mean that the cluster is either not installed or in trouble, but I think that's a fine trade-off for now. Gravity status also does it this way currently.

Reply:
> For example, the way you mentioned also won't work for scenarios where, say, a cluster was partially installed/uninstalled and still has some leftovers on disk.

I'm not sure what we should consider "working" in this particular scenario. Should pre-install checks be run and give a clean bill of health, for the installer to detect the install state and barf later?

I think based on this discussion, there are 4 scenarios, and the way it works now:
1. Clean system - pre-install checks
2. Partial install / Non working system - pre-install checks
3. Partial install - planet-agent online - upgrade checks (could report system as healthy or non healthy depending on where the install failed, unless we set something like install state in etcd, I've never checked that)
4. completed install - upgrade checks

Using some level of installation state on disk
1. Clean system - pre-install checks
2. Partial install / Non working system - report cluster as non-functioning
3. Partial install - planet-agent online - could report healthy/non healthy
4. completed install - upgrade checks

The installer errors are here:
```
[ERROR]: detected previous installation state in /var/lib/gravity/local, please clean it up using `gravity leave --force` before proceeding (see https://gravitational.com/gravity/docs/cluster/#deleting-a-cluster for more details)
```
https://github.com/gravitational/gravity/blob/master/lib/system/environ/validate.go#L67-L71

https://github.com/gravitational/gravity/blob/master/lib/system/environ/validate.go#L136-L141


Perhaps the issue is that we're continuing to overload logic that operates in non-intuitive ways. The pre-install and pre-upgrade checks are similar but not identical, but we're trying to create a single UX pattern for both which will confuse those who encounter it. Because it has to guess what the user intends to do, and there is always potential for that guess to be wrong. The `gravity status` behavior to report pre-install checks when it can't connect to the cluster is already extremely confusing to those who encounter it. 


I dunno, to me it makes more sense to report the system as non functioning on a partial install, than it does to give it a clean bill of health only to fail when someone tries to run the install a moment later.

Reply:
@knisbet Yeah, I think it would make sense to detect partial installation and give the instructions  on how to clean up before running the checks. I've already merged this but I'll make a ticket to improve this behavior.